pyam/core.py

```diff
@@ -722,118 +716,140 @@ def check_aggregate(self, variable, components=None, units=None,
         """
         # compute aggregate from components, return None if no components
-        df_components = self.aggregate(variable, components, units)
+        df_components = self.aggregate(variable, components, unit)
         if df_components is None:
             return
 
         # filter and groupby data, use `pd.Series.align` for matching index
+        rows = self._apply_filters(variable=variable, unit=unit)
         df_variable, df_components = (
-            _aggregate_by_variables(self.data, variable, units)
-            .align(df_components)
+            _aggregate(self.data[rows], 'variable').align(df_components)
         )
 
         # use `np.isclose` for checking match
         diff = df_variable[~np.isclose(df_variable, multiplier * df_components,
                                        **kwargs)]
 
         if len(diff):
-            msg = '{} - {} of {} data points are not aggregates of components'
+            msg = '`{}` - {} of {} rows are not aggregates of components'
             logger().info(msg.format(variable, len(diff), len(df_variable)))
 
             if exclude_on_fail:
                 self._exclude_on_fail(diff.index.droplevel([2, 3, 4]))
 
-            diff = pd.concat([diff], keys=[variable], names=['variable'])
+            return IamDataFrame(diff, variable=variable).timeseries()
+
+    def aggregate_region(self, variable, region='World', subregions=None,
+                         components=None, unit=None, append=False):
+        """Compute the aggregate of timeseries over a number of regions
+        including variable components only defined at the `region` level
 
-            return diff.unstack().rename_axis(None, axis=1)
+        Parameters
+        ----------
+        variable: str
+            variable for which the aggregate should be computed
+        region: str, default 'World'
+            dimension
+        subregions: list of str
+            list of subregions, defaults to all regions other than `region`
+        components: list of str
+            list of variables, defaults to all sub-categories of `variable`
+            included in `region` but not in any of `subregions`
+        unit: str or list of str
+            filter variable and variable components for given unit(s)
+        append: bool, default False
+            append the aggregate timeseries to `data` and return None,
+            else return aggregate timeseries
+        """
+        # default subregions to all regions other than `region`
+        if subregions is None:
+            rows = self._apply_filters(variable=variable)
+            subregions = set(self.data[rows].region) - set([region])
 
-    def check_aggregate_regions(self, variable, region='World',
-                                components=None, units=None,
-                                exclude_on_fail=False, **kwargs):
+        if not len(subregions):
+            msg = 'cannot aggregate variable `{}` to `{}` because it does not'\
```

Comment:
is this message right @danielhuppmann ?

Reply:
I'd say yes: if the variable `Emissions|CO2` is not defined in any subregion, the function should not make any assumption that this is zero at the `World` level. And if you want to add up subcategories of `Emissions|CO2|*` at the World level, you should use `aggregate()` rather than `aggregate_region()`.